src/io/ReadHandle.php

```diff
@@ -34,8 +34,25 @@ public function rawReadBlocking(?int $max_bytes = null): string;
    * end of file, or the timeout is reached.
    *
    * If `$max_bytes` is 0, the empty string will be returned.
+   *
+   * @see readPartialAsync
+   * @see readLineAsync
+   */
+  public function readAllAsync(
+    ?int $max_bytes = null,
+    ?float $timeout_seconds = null,
+  ): Awaitable<string>;
+
+  /** Read the next readable chunk.
+   *
+   * This will wait until end of file or the timeout is reached; it will
```

Comment:
> Is this correct? 

No, it will wait until end of file, timeout, or there is data available to read.

> Also, this might be a limitation of my network knowledge, but is "no more data available to read" a well defined concept?

the state when a non-blocking read will return data; the connection could be open but idle. It does not require that *all* the data is available to read.

Reply:
Thanks - I'll be updating these tomorrow. Just trying to wrap stuff up for my work in progress :)

Reply:
There's a few distinct concepts for end of file, and AIUI the current behavior of `\feof()` is not entirely consistent across various resource types

1. there is currently no more data available to read, but there may be later - e.g. a blocking `read()` will return 0 (bytes read), and a non-blocking `read()` will fail with `EAGAIN`; this might be a socket or a pipe that is still open, but the write end hasn't sent anything yet
2. there will never be more data to read - this might mean you've reached the end of a physical file (`read()` will return 0) or various error conditions that can be considered similar to "it's a socket closed by the other end" - ENOTCONN, ECONNRESET, ESTALE (NFS server gone away) etc

libc uses end of file to cover all of these - just "no data right now".

I think we should either:
- use it to match libc consistently
- move it to FileHandle; the other states can be detected by read attempts returning 0 bytes vs throwing exceptions

Reply:
@wez any thoughts on handling/defining EOF?

Reply:
re: EOF, it can vary depending on the underlying fd.  For example, I recently learned that EOF for the master side of a pty indicates EOF by returning an EIO error when the slave side of the pty is closed.   For pipes you might get EPIPE in that situation.  For files you'll just get a 0 byte successful read and so on.

I've come to have the opinion that we don't need `feof()` as a concept.  It is present in PHP because it was present in the ANSI FILE* stuff and that is really an artifact of how it does buffering.

I would lean instead towards just having folks repeat read/write operations and deal with the error condition appropriate if they encounter one.  And *maybe* (but probably not) buffer the last error if we're sure that it will be a persistent error, but that would just be an optimization for an uncommon error path.

I know that backwards compatibility may make it difficult to be this opinionated here :)

Reply:
Especially in hsl-experimental, happy to do 'the best thing' rather than the most-BC thing :)